pkg/reconciler/instances/istio/webhooks/webhooks_test.go

```diff
@@ -0,0 +1,125 @@
+	}},
+}
+
+var deactivatedSelector = &metav1.LabelSelector{
+	MatchLabels: deactivatedLabel,
+}
+
+func createDefaultMutatingWebhookWithSelector(whConfName string, labelKey string, selector *metav1.LabelSelector) *v1.MutatingWebhookConfiguration {
+	return &v1.MutatingWebhookConfiguration{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   whConfName,
+			Labels: map[string]string{labelKey: tag.DefaultRevisionName},
+		},
+		Webhooks: []v1.MutatingWebhook{
+			{
+				Name:              "object.sidecar-injector.istio.io",
+				NamespaceSelector: selector,
+			},
+		},
+	}
+}
+
+func Test_DeleteConflictedDefaultTag(t *testing.T) {
+	kubeConfig := "kubeconfig"
+	log := logger.NewLogger(false)
+	ctx := context.Background()
+	defaultWhName := "istio-sidecar-injector"
+	taggedWhName := "istio-revision-tag-default"
+
+	defer ctx.Done()
+
+	t.Run("should return error when kubeclient could not be retrieved", func(t *testing.T) {
+		// given
+		provider := clientsetmocks.Provider{}
+		provider.On("RetrieveFrom", mock.AnythingOfType("string"), mock.AnythingOfType("*zap.SugaredLogger")).Return(nil, errors.New("Kubernetes client error"))
+
+		// when
+		err := DeleteConflictedDefaultTag(ctx, &provider, kubeConfig, log)
+
+		// then
+		require.Error(t, err)
+		require.Contains(t, err.Error(), "Kubernetes client error")
+	})
+	t.Run("should not delete tagged webhook when old webhook is deactivated", func(t *testing.T) {
+		// given
+		defaultWh := createDefaultMutatingWebhookWithSelector(defaultWhName, "istio.io/rev", deactivatedSelector)
+		taggedWh := createDefaultMutatingWebhookWithSelector(taggedWhName, tag.IstioTagLabel, validSelector)
+		client := fake.NewSimpleClientset(defaultWh, taggedWh)
+		provider := &clientsetmocks.Provider{}
+		provider.On("RetrieveFrom", mock.AnythingOfType("string"), mock.AnythingOfType("*zap.SugaredLogger")).Return(client, nil)
+		// when
+		err := DeleteConflictedDefaultTag(ctx, provider, kubeConfig, log)
+
+		// then
+		require.NoError(t, err)
+		got, err := client.AdmissionregistrationV1().MutatingWebhookConfigurations().Get(ctx, taggedWhName, metav1.GetOptions{})
+		require.NoError(t, err)
+		require.Equal(t, taggedWh, got)
+	})
+	t.Run("should delete conflicted tagged webhook if old one is not deactivated", func(t *testing.T) {
+		// given
+		defaultWh := createDefaultMutatingWebhookWithSelector(defaultWhName, "istio.io/rev", validSelector)
+		taggedWh := createDefaultMutatingWebhookWithSelector(taggedWhName, tag.IstioTagLabel, validSelector)
+		client := fake.NewSimpleClientset(defaultWh, taggedWh)
+		provider := &clientsetmocks.Provider{}
+		provider.On("RetrieveFrom", mock.AnythingOfType("string"), mock.AnythingOfType("*zap.SugaredLogger")).Return(client, nil)
+		// when
+		err := DeleteConflictedDefaultTag(ctx, provider, kubeConfig, log)
+
+		// then
+		require.NoError(t, err)
+		got, err := client.AdmissionregistrationV1().MutatingWebhookConfigurations().Get(ctx, taggedWhName, metav1.GetOptions{})
```

Comment:
Adding an assert that the default webhook is not deleted might be a good idea. What do you think?